[ISSUE-5705] Support for degraded health

seonWKim opened the pull request:
Motivation:

Add support for degraded health. 

Modifications:

- Add `HealthStatus` enum which can represent the health status of a server 
- Add `healthStatus()` method in `HealthChecker` for fine-grained representation of server's status 
- Add `degradedResponse` which shows that the server's status is being degraded. 

I think adding new health check service which returns `HealthStatus`(instead of boolean) might be an alternate option. But because this PR is before any reviews, I've just implemented feature on top of `HealthCheckService`.   
 
Result:

Users can now receive notification for degraded server status like below 
```json 
{"healthy":true, "degraded":true}
``` 

- Closes <https://github.com/line/armeria/issues/5705>. (If this resolves the issue.)
- Users can now identify `degraded` server status in addtion to healthy and unhealthy.

<!--
Visit this URL to learn more about how to write a pull request description:
https://armeria.dev/community/developer-guide#how-to-write-pull-request-description
-->

seonWKim committed:
Add support for degradable health status in `HttpCheckService`

seonWKim committed:
Fix `CpuHealthChecker` and add test

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java:
```diff
@@ -295,14 +294,14 @@ public void serverStarting(Server server) throws Exception {
             public void serverStopping(Server server) {
                 serverStopping = true;
-                serverHealth.setHealthy(false);
+                serverHealth.setHealthStatus(HealthStatus.UNHEALTHY);
```
Should we also introduce `HealthStatus.UNDER_MAINTENANCE`?

seonWKim replied:
For finer grained control, it seems a great idea. 

What do you think of the below responses for each status?(I'm a bit worried if this will break backward compatibility though)  
- HEALTHY: `{"healthy":true, "status": "HEALTHY"}` 
- DEGRADED: {"healthy":true, "status": "DEGRADED"}` 
- UNDER_MAINTENANCE: {"healthy":true, "status": "UNDER_MAINTENANCE"}`  
- UNHEALTHY: {"healthy":false, "status": "UNHEALTHY"}`

seonWKim replied:
Sure, I'll add it. 👍

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckServiceBuilder.java:
```diff
@@ -306,7 +331,7 @@ public HealthCheckService build() {
                    "Healthiness must be updatable by server listener or update handler.");
         return new HealthCheckService(healthCheckers.build(),
-                                      healthyResponse, unhealthyResponse,
+                                      healthyResponse, degradedResponse, unhealthyResponse,
```
We could also add `underMaintenanceResponse`?

seonWKim replied:
Seems like the number of arguments for the responses increases(`healthyResposne`, `unhealthyResponse`, `underMaintenanceResponse`, `degradedResponse`). Should we receive the responses as `Collections` instead of  adding it as an argument? This way, we won't have to add another argument when another `HealthStatus` is added.

seonWKim replied:
Yes, we can. What do you think of the below responses for each responses?(I'm a bit worried if this will break backward compatibility though)

`healthyResponse`: `{"healthy":true, "status": "HEALTHY"}`
`unhealthyResponse`: `{"healthy":true, "status": "DEGRADED"}`
`degradedResponse`: `{"healthy":true, "status": "UNDER_MAINTENANCE"}`
`unhealthyResponse`: `{"healthy":false, "status": "UNHEALTHY"}`

trustin replied:
I'd prefer adding `status` for all responses.

seonWKim replied:
I've added status field to all 5 responses 
- healthyResponse 
- degradedResponse
- serverStoppingResponse 
- underMaintenanceResponse
- unhealthyMaintenance

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthStatus.java:
```diff
@@ -0,0 +1,50 @@
+        return priority;
+    }
+
+    /**
+     * Returns whether this {@link HealthStatus} is considered healthy.
+     */
+    public boolean isHealthy() {
+        return isHealthy;
+    }
```
Maybe this should be renamed to `isAvailable()` or something else? I find it not intuitive to call 'degraded' state as 'healthy'.

seonWKim replied:
I agree `isAvailable()` better describes the condition.

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/SettableHealthStatusChecker.java:
```diff
@@ -0,0 +1,83 @@
+ * health should have additional conditions besides the state of the {@link Server}. e.g. it should depend on
+ * the health of a backend.
+ */
+public final class SettableHealthStatusChecker extends AbstractListenable<HealthChecker>
```
- Can we merge this into `SettableHealthChecker`? Is there any reason we have to keep both `SettableHealthChecker` and `SettableHealthStatusChecker`?
- There are no tests for this class.

seonWKim replied:
>Can we merge this into SettableHealthChecker? Is there any reason we have to keep both SettableHealthChecker and SettableHealthStatusChecker?

Yes, we can merge it to `SettableHealthChecker`. I separated the classes because I thought the code would be too messy to maintain backward compatibility of `SettableHealthChecker` and add new feature(setting `HealthStatus`). However, I think it's possible to maintain both backward compatibility and adding new feature without making `SettableHealthChecker` too messy. 

>There are no tests for this class.

I'm gonna remove `SettableHealthStatusChecker`, so I'll add some test into `SettableHealthChecker`. Thank you for the through review.

trustin replied:
Thanks!

trustin reviewed (commented):
Thanks, @seonWKim! :bow:

seonWKim committed:
Remove `SettableHealthStatusChecker` and merge the features to `SettableHealthChecker`

seonWKim committed:
Add `healthStatusUpdated` for more fine fine-grained health status updates

seonWKim committed:
Add `UNDER_MAINTENANCE`

seonWKim commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckServiceBuilder.java:
```diff
@@ -50,9 +50,19 @@ public final class HealthCheckServiceBuilder implements TransientServiceBuilder
     private AggregatedHttpResponse healthyResponse = AggregatedHttpResponse.of(HttpStatus.OK,
                                                                                MediaType.JSON_UTF_8,
                                                                                "{\"healthy\":true}");
```
Would it break backward compatibility if we use `"{\"healthy\":true,\"status\":\"HEALTHY\"}"` for `healthyResponse`?

trustin replied:
I think it's OK adding a new field to a scheme isn't a breaking change in JSON world.

seonWKim replied:
I've added status field 🙇

seonWKim committed:
Fix getResponse

seonWKim commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java:
```diff
@@ -517,40 +524,48 @@ private HttpResponse newResponse(HttpMethod method, boolean isHealthy) {
 
-    private AggregatedHttpResponse getResponse(boolean isHealthy) {
-        if (isHealthy) {
+    private AggregatedHttpResponse getResponse(HealthStatus healthStatus) {
```
What if we add additional `SERVER_STOPPING` status. That way, we can make `getResponse` method much clear.

trustin replied:
I kind of though `UNDER_MAINTENANCE` is a generalized version of that. What do you think about adding it once we really have a use case for it?

seonWKim replied:
Sure, let's add it! 😄

seonWKim committed:
Fix lint

seonWKim committed:
Fix test name

seonWKim committed:
Fix docs

seonWKim committed:
Add (server) STOPPING status

seonWKim committed:
Fix test

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java:
```diff
@@ -323,15 +322,20 @@ public void serverStopped(Server server) throws Exception {
                     Ascii.toLowerCase(req.headers().get(HttpHeaderNames.IF_NONE_MATCH, ""));
             if ("\"healthy\"".equals(expectedState) || "w/\"healthy\"".equals(expectedState)) {
-                useLongPolling = isHealthy;
+                useLongPolling = healthStatus == HealthStatus.HEALTHY;
```
note: this change means if `HEALTHY` or `UNHEALTHY` isn't set, then long polling never works since armeria client only sends `healthy` or `unhealthy` depending on the status code.

https://github.com/line/armeria/blob/579a6d0998e34605c41ea13b26c1f7c19836e908/core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HttpHealthChecker.java#L105-L106

seonWKim replied:
Oh, I see. Than I think we can change the logic as follows:

When the `expectedState` is
- `healthy`:  use long polling when the status is `HEALTHY` or `DEGRADED` 
- `unhealthy`:  use long polling when the status is `UNHEALTHY` or `UNDER_MAINTENANCE`

jrhee17 replied:
I understood the objective of this issue to propagate the status to clients.
If we only maintain `expectedState`, I think clients may not receive the status on update for long polling health check requests.

If we would like to maintain backwards compatibility, we could define a separate syntax for status matching.

e.g.
- `If-None-Match=[healthy|unhealthy]`: propagates a signal whenever a status changes between healthy <-> unhealthy
- `If-None-Match=status=[HEALTHY|DEGRADED|UNHEALTHY|UNDER_MAINTENANCE|STOPPING]`: propagates a signal when the status does not match the status exactly

seonWKim replied:
I've added `If-None-Match=status=[HEALTHY|DEGRADED|UNHEALTHY|UNDER_MAINTENANCE|STOPPING]` and preserved `If-None-Match=[healthy|unhealthy]` in order to maintain backward compatibility. 

- When `If-None-Match=[healthy|unhealthy]` is present, return a response when the server becomes `healthy` or `unhealthy` 
- When `If-None-Match=status=[HEALTHY|DEGRADED|UNHEALTHY|UNDER_MAINTENANCE|STOPPING]` is present, return a response when the server's `HealthStatus` is changed.

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java:
```diff
@@ -435,24 +435,35 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
                 switch (updateResult) {
                     case HEALTHY:
-                        serverHealth.setHealthy(true);
+                        serverHealth.setHealthStatus(HealthStatus.HEALTHY);
```
Note: If the default `HealthCheckUpdateHandler` is used, only `HEALTHY` and `UNHEALTHY` will be returned. 

I think users would find it surprising that the server is aware of options such as `DEGRADED`, etc.. but doesn't understand update semantics of such statuses

seonWKim replied:
Oh you're right. What do you think of adding an implementation of `HealthCheckUpdateHandler` to handle new status? 🤔

jrhee17 replied:
Sounds good to me 👍

seonWKim replied:
Added `HealthStatusUpdateHandler` 😄

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckServiceBuilder.java:
```diff
@@ -105,6 +119,44 @@ public HealthCheckServiceBuilder healthyResponse(AggregatedHttpResponse healthyR
+     *
+     * @return {@code this}
+     */
+    public HealthCheckServiceBuilder degradedResponse(AggregatedHttpResponse degradedResponse) {
```
note: from the perspective of health checking in envoy, just sending a `x-envoy-degraded` header to denote an endpoint is probably enough. Checked that this won't interfere with ongoing xDS related changes.

seonWKim replied:
It seems envoy use headers to denote the specific status. What would be more appropriate for armeria to denote specific status? The headers or status field in the json?

jrhee17 replied:
I think it's fine to let users customize the response for now - later depending on requirements we may just introduce a separate `EnvoyHealthCheckService`.

trustin replied:
We probably should do some research on the behavior of well known server implementations and make the default response covers them all, as a follow-up to this PR.

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckServiceBuilder.java:
```diff
@@ -47,12 +47,26 @@ public final class HealthCheckServiceBuilder implements TransientServiceBuilder
-                                                                                 "{\"healthy\":false}");
+    private AggregatedHttpResponse healthyResponse =
+            AggregatedHttpResponse.of(HttpStatus.OK, MediaType.JSON_UTF_8,
+                                      "{\"healthy\":true,\"status\":\"HEALTHY\"}");
```
Note: In case this point was missed, armeria client is ignoring the body at the moment and determining health status via status code.

seonWKim replied:
Oh right. `HttpHealthChecker` [only checks the status code](https://github.com/line/armeria/blob/fea0622618613631735b4625c3bf8faf20d9f87a/core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/HttpHealthChecker.java#L211). Should we handle the changes in this PR as well?

jrhee17 replied:
I'm unsure of how we want to expose health check information on the client-side at the moment. From an xDS perspective, other signals such as 1) the ratio of failed health checks 2) time since last health check 3) etc.. are exposed.

I think it's better to focus on implementing backwards compatible server-side changes in this PR, and client-side changes are dealt with separately (after a consensus is made on the API direction).

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java:
```diff
@@ -178,15 +179,13 @@ public static HealthCheckServiceBuilder builder() {
             healthCheckerListener = this::onHealthCheckerUpdate;
-            pendingHealthyResponses = new ObjectLinkedOpenHashSet<>();
-            pendingUnhealthyResponses = new ObjectLinkedOpenHashSet<>();
+            pendingResponses = ConcurrentHashMap.newKeySet();
```
nit; `pendingResponses` is always updated inside a lock already

seonWKim replied:
You're right. It seems we can still use `ObjectLinkedOpenHashSet` as the lock is already used.

jrhee17 reviewed (commented):
Looks good overall 👍   I think some client-side changes will be needed, but I prefer I do this to ensure compatibility with xDS-side requirements. Left some minor comments 🙇

seonWKim committed:
Apply suggestions from code review

Co-authored-by: jrhee17 <guins_j@guins.org>

seonWKim committed:
Apply comments

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthStatus.java:
```diff
@@ -0,0 +1,54 @@
+ */
+@UnstableApi
+public enum HealthStatus {
+    HEALTHY(500, true),
+    DEGRADED(400, true),
+    STOPPING(300, false),
+    UNHEALTHY(200, false),
+    UNDER_MAINTENANCE(100, false);
```
Javadoc is missing.

seonWKim replied:
Added javadoc 👍

trustin commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthStatus.java:
```diff
@@ -0,0 +1,54 @@
+    /**
+     * Returns the priority of this {@link HealthStatus}.
+     */
+    public int getPriority() {
```
```suggestion
    public int priority() {
```

seonWKim replied:
Fixed! Thanks for the detailed review 🙇

trustin reviewed (commented):
Almost there!

seonWKim committed:
Fix to code to consider backward compatibility and allow users to specify HealthStatus

seonWKim committed:
Apply comments
- Add javadoc on `HealthStatus`
- Checks for IF-NONE-MATCH header for specific `HealthStatus`
- Add `HealthStatusUpdateHandler`

seonWKim commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/DefaultHealthCheckUpdateHandler.java:
```diff
@@ -32,9 +32,14 @@
+ * Handler which updates the healthiness of the {@link Server}. Supports {@code PUT}, {@code POST} and
+ * {@code PATCH} requests and tells if the {@link Server} needs to be marked as healthy or unhealthy.
+ */
+public enum DefaultHealthCheckUpdateHandler implements HealthCheckUpdateHandler {
```
I changed the access modifier to public in order to let users select their update handler`(`DefaultHealthCheckUpdateHandler` or `HealthStatusUpdateHandler`)

jrhee17 replied:
Question) what do you think of just combining the two `HealthCheckUpdateHandler` and adding similar logic as `HealthCheckService`?

e.g.
```
if jsonNode.contains("healthy") {
  // previous logic
} else if jsonNode.contains("status") {
  // update to the status
} else {
  // 400
}
```

trustin replied:
Once done, please don't forget to make this class package-private again.

seonWKim committed:
Set `HealthStatusUpdateHAndler` to public

seonWKim commented:
Removed the `[ISSUE-5741]` prefix in the commit messages.

jrhee17 reviewed (commented):
Looks good overall 👍  Left a minor suggestion

seonWKim commented:
>Question) what do you think of just combining the two HealthCheckUpdateHandler and adding similar logic as HealthCheckService?

Looks better. I don't think we need to split them into different classes. 👍

seonWKim committed:
Merge `HealthStatusUpdateHandler` features into `DefaultHealthCheckUpdateHandler`

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/SettableHealthChecker.java:
```diff
@@ -46,22 +44,45 @@ public SettableHealthChecker() {
      */
     public SettableHealthChecker(boolean isHealthy) {
-        this.isHealthy = new AtomicBoolean(isHealthy);
+        this.healthStatus = isHealthy ? HealthStatus.HEALTHY : HealthStatus.UNHEALTHY;
```
```suggestion
        healthStatus = isHealthy ? HealthStatus.HEALTHY : HealthStatus.UNHEALTHY;
```

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthChecker.java:
```diff
@@ -105,8 +106,37 @@ static HealthChecker ofCpu(double targetSystemCpuUsage, double targetProcessCpuU
+     * Returns the {@link HttpStatus} representing the health status of the {@link Server}.
+     * Override below method if you want more fine-grained health status.
+     */
+    default HealthStatus healthStatus() {
```
Can you modify that `ScheduledHealthChecker` also uses the `healthStatus` API?

seonWKim replied:
Updated! 👍

jrhee17 commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthChecker.java:
```diff
@@ -105,8 +106,37 @@ static HealthChecker ofCpu(double targetSystemCpuUsage, double targetProcessCpuU
     /**
      * Returns {@code true} if and only if the {@link Server} is healthy.
      */
     boolean isHealthy();
```
Note: I imagine that some time in the future we may likely end up deprecating this API since we're only using `HealthChecker#healthStatus` internally in `HealthCheckService`.

I prefer this be handled separately though if necessary.

jrhee17 replied:
Meanwhile, can you go through the code in armeria and convert any usages using this API to use `healthStatus` instead??

seonWKim replied:
I've updated the code to use `healthStatus()` instead of using `isHealthy()` API. The only place where `isHealthy()` is used is below method. 

```java 
default HealthStatus healthStatus() {
    return isHealthy() ? HealthStatus.HEALTHY : HealthStatus.UNHEALTHY;
}
```

jrhee17 reviewed (commented):
This is probably the last of my comments.
Please also address https://github.com/line/armeria/pull/5741/files#r1651918233 🙏

seonWKim committed:
Apply comments
- Update access modifier of `DefaultHealthCheckUpdateHandler`
- Update `ScheduledHealthChecker` to use healthStatus
- Remove unnecessary `this`

jrhee17 reviewed (commented):
Left some more nit comments 👍

seonWKim committed:
Apply comments
- Add @UnstableApi

seonWKim commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/ScheduledHealthChecker.java:
```diff
@@ -55,16 +54,22 @@ final class ScheduledHealthChecker extends AbstractListenable<HealthChecker>
+
+    @Override
+    public HealthStatus healthStatus() {
+        return healthStatusAtomicReference.get();
```
Yes. Missed that. I've modified `HealthCheckStatus` to accept `HealthStatus` as well.

seonWKim committed:
Modify `HealthCHeckStatus` to accept `boolean` and `HealthStatus`

jrhee17 reviewed (approved):
👍 👍 👍

minwoox commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckUpdateResult.java:
```diff
@@ -25,10 +25,22 @@ public enum HealthCheckUpdateResult {
+    /**
+     * Tells {@link HealthCheckService} to mark the {@link Server} as 'under maintenance'.
+     */
+    UNDER_MAINTENANCE,
```
Can we just use MAINTENANCE?

minwoox committed:
Merge branch 'main' into pr-5741-seonWKim-feature/ISSUE-5705-degradable-health-checker

minwoox committed:
More fix

minwoox commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthStatus.java:
```diff
@@ -0,0 +1,72 @@
+    /**
+     * The {@link Server} is under maintenance and unable to serve requests.
+     */
+    UNDER_MAINTENANCE(100, false);
```
Better to put `UNDER_MAINTENANCE` after `STOPPING`? We also need to update the Javadoc of `priority` to explain the value means.

minwoox commented on core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java:
```diff
@@ -517,40 +540,54 @@ private HttpResponse newResponse(HttpMethod method, boolean isHealthy) {
+                    this.pendingResponses.stream()
+                                         .filter(res -> {
+                                             if (res.pollHealthStatus) {
+                                                 return res.interestedStatus != healthStatus;
```
Shouldn't it filter when
- the `healthStatus` is changed from its previous `healthStatus`
- `res.interestedStatus == healthStatus`;

minwoox reviewed (commented):
Hi, @seonWKim!
Sorry about the late review.
I've pushed some fixes and left some comments to discuss. 😉

seonWKim commented:
I completely missed this, greatly sorry to all the reviewers. If this has not been handled yet, I will take care of it immediately.
@minwoox, is this issue still open? I’ll work on it as quickly as possible.
